io-engine/src/rebuild/rebuild_log.rs

```diff
@@ -109,43 +84,58 @@ impl RebuildLog {
         (lbn * self.block_len / SEGMENT_SIZE) as usize
     }
 
-    /// Counts the total number of modified blocks.
-    pub fn count_modified_blocks(&self) -> u64 {
-        let segs = self.segments.iter().filter(|i| *i).count() as u64;
-        segs * SEGMENT_SIZE / self.block_len
+    /// Counts the total number of blocks set to one.
+    pub fn count_ones(&self) -> u64 {
+        self.segments.iter().filter(|i| *i).count() as u64
+    }
+
+    /// Counts the total number of dirty blocks.
+    pub fn count_dirty_blks(&self) -> u64 {
+        self.count_ones() * SEGMENT_SIZE / self.block_len
     }
 }
 
-/// A handle object for a rebuild log instance.
-/// A rebuild log instance allows to have multiple handle to access it.
-#[derive(Clone)]
-pub struct RebuildLogHandle {
-    /// Share log reference.
-    log: Arc<Mutex<RebuildLog>>,
-    /// Name of the underlying device.
-    /// Keep it here to avoid locking just to access the device name,
-    /// and cloning the name when accessing it.
+/// Per-core I/O log channel.
+/// I/O log channel is enables lockless logging of I/O operations.
+pub struct IOLogChannelInner {
+    /// Channel's core.
+    core: u32,
+    /// Name of the underlying block device.
     device_name: String,
+    /// Map of device segments.
+    segments: UnsafeCell<Option<SegmentMap>>,
```

Comment:
Don't we have mutable access to the channel? Then we can make log_io take &mut self, and not have to use `UnsafeCell` ?

Reply:
IoLogChannel is accessed via Rc<>, which would require an UnsafeCell here or as Rc<UnsafeCell<>>. I prefer it here, as device_name is logically const.